use-cases/deep-replace.md

```diff
@@ -0,0 +1,78 @@
+# Deeply Replacing Packages in Nixpkgs
+
+## General Information
+
+Replace all "instances" of a package across a dependency graph.
+This may be replacing all usage in a package set, collection of package sets,
+a collection of ad-hoc recipes, or a collection of flakes.
+
+A possible motivation for _deep replacement_ may be to ensure that a security
+fix provided by a new release of a piece of software is used "everywhere" in
+the dependency graph.
+
+
+## Concrete Examples
+
+While the precise organization of packages will effect the complexity and
+effort required to perform _deep replacement_, in general we say that this
+is accomplished using helper functions such as `extend`, `overrideScope'`, and
+`appendOverlays`, as well as the configuration field `overlays`.
+
+### Simple Overlay
+
+```nix
+pkgs.extend ( final: prev: { foo = final.callPackage ./my-pkgs/foo {}; } )
+```
+
+
+<a name="Nested-Ex"></a>
+### Nested Overlays
+
+This fails to use Node.js v14 in `nodePackages`, which is unexpected to
+most users.
+```nix
+# XXX: Incorrect usage
+pkgs.extend ( final: prev: {
+  nodejs       = prev.nodejs-14_x;
+  nodePackages = prev.nodePackages.extend ( nfinal: nprev: {
+    bar = nfinal.callPackage ./my-pkgs/bar {};
+  } );
```

Comment:
Great example!

Reply:
Wait a second, why does this not work? `prev.nodejs` isn't referenced anywhere in this overlay, so the old value shouldn't be used anywhere either, right?

Reply:
I think it's because `prev.nodePackages.extend` is used instead of `final.nodePackages.extend`

Reply:
I just wanted to try this, out but turns out `nodePackages.callPackage` doesn't exist, so this code doesn't work:
```
error: attribute 'callPackage' missing

       at /home/tweagysil/src/nixpkgs/ext.nix:6:15:

            5|       nodePackages = prev.nodePackages.extend ( nfinal: nprev: {
            6|         bar = nfinal.callPackage ./my-pkgs/bar {};
             |               ^
            7|       } );
```

@aakropotkin Can you give an example that can be reproduced?

Reply:
yes absolutely

Reply:
Still, I think the mistake here is using:
```
nodePackages = prev.nodePackages.extend
```
instead of
```
nodePackages = curr.nodePackages.extend
```

If that is correct, then the example doesn't demonstrate a problem with deeply replacing packages, but a bad UX of the override concept, that allows the user to confuse `prev` and `curr`.